Add bytes and string serializers

bbenne10 opened the pull request:
Begins to address #6.
Some family matters will keep me from revisiting this this weekend, but this builds locally.
I'll still need to add documentation and tests for the added functionality.

Mostly opening early for feedback on the implementation.
I've chosen to go with `Buffers` internally, as they are relatively fast (from what I have read) and easily convert between bytes and strings.
The imperative nature of working with them is frustrating though.
This can be refactored pretty easily with a `String` based interface, if desired.

You'll notice that I didn't use `Bitv.output_bin`.
It only accepts channels and creating a memory-backed channel seemed quite daunting.
Instead, I just used `Bitv.to_list` and re-used some of the logic that {de,}serializes `p_len`.

bbenne10 committed:
Add bytes and string serializers

bbenne10 commented:
Is the failing stage a lack of changes to `CHANGES.md`?
I can add those when I get back to working on this. 
If someone cares to explain the process (if there is one), I can also bump the version if desired.

bbenne10 commented:
I noticed that I'd forgotten to `ocamlformat` the interface file, so I've force-pushed a fix for that.
The other failure looked to be a cancelled build somewhere? 
Will revisit when I come back to this more seriously :)

bbenne10 commented:
Ah - the failure is building under OCaml 4.03, as I'm using `String.split_on_char` which was introduced in 4.04.
I'm unsure how to resolve that (or if I need to - Does OCaml have a deprecation policy for OCaml versions like Python does? If so - is 4.03 still supported? Where might I find out? Initial googling doesn't say much)

It looks simple enough to copy the stdlib implementation into our code, if that is required?

bbenne10 commented:
Serializing to strings seems to work, but there's a failure during parsing which I'm tracking down now.

If anyone gets around to it and notices where I've screwed up - please let me know.
If not - a general critique of my approach would be appreciated.

Edited to add: I also note that we rely on `Random.self_init`. Is there a reason for this? Because of this, I cannot reliably test that `to_string` returns a known good value. I can only test that it can read the values that it writes.

dinosaure commented:
As a MirageOS project, we prefer to use the `result` type (instead `failwith`) to inform about an error while parsing (see this article about [error handling in MirageOS](https://mirage.io/wiki/mirage-3.0-errors)) - a quick look on the PR tell me a leak of exception with `int_of_string`. About testing, `alcotest` is a good library to implement unit tests. However, for such feature, it can be interesting to use [crowbar](https://github.com/stedolan/crowbar) to randomly test your functions and assert some assumptions (such as `x = of_string (to_string x))`).

bbenne10 commented:
Frankly, I prefer the `result` anyway. It makes the code longer internally, but I think that's fine when presenting a better interface. 

I'm fighting with the exception leak over `int_of_string` right now (its what I mentioned as the problem earlier).
If I rework to use `result`, this specific problem will disappear, but it indicates that there's a much bigger problem (as it's failing to read it's own written data!)

Thanks for your feedback.
I'll work towards using `result` and resolving the issue noted with `int_of_string`

(Really missing `int_of_string_opt` in 4.04 and lower right about now :P)

pascutto commented on src/bloomf.mli:
```diff
@@ -36,6 +36,14 @@ val size_estimate : 'a t -> int
+val of_string : string -> 'a t
+
+val of_bytes : bytes -> 'a t
+
```
Could you please add some documentation here?

pascutto commented on src/bloomf.ml:
```diff
@@ -78,6 +78,75 @@ let size_estimate t =
+    let buf = Buffer.create (2 * List.length bitv) in
+    bitv |> List.iter (fun b -> string_of_int b ^ rs |> Buffer.add_string buf);
+    buf |> Buffer.contents
+  in
```
This function encodes the bit vector as the list of the indices of the bits set to `1`. That takes considerably more space than the bit vector itself if its fill rate is too high. Have you measured that effect somehow, to ensure it's not getting too big?
I would suggest using the [import/export functions provided by `Bitv`](https://github.com/backtracking/bitv/blob/master/bitv.mli#L161-L188) instead, as it will also make this code simpler and less error-prone.

pascutto commented on src/bloomf.ml:
```diff
@@ -78,6 +78,75 @@ let size_estimate t =
+        bv_str
+        |> String.split_on_char rs
+        |> List.map int_of_string
+        |> Bitv.of_list
```
I think there might be a mistake here, as `Bitv.of_list` won't know about the original size of the vector, but will instead deduce it from the max element of the list (ie, the last bit set to `1`), which is probably not what you want. 
You would also avoid that caveat by using the functions provided by `Bitv`.

bbenne10 replied:
You're right. I'll switch to `L.to_string` and `L.of_string`. I can't believe I missed those when looking for alternatives to `{input,output}_bin`...

pascutto commented on src/bloomf.ml:
```diff
@@ -78,6 +78,75 @@ let size_estimate t =
+      in
+      let p_len = parse_plen p_len_str in
+      { m = int_of_string m_str; k = int_of_string k_str; p_len; b }
+  | _ -> failwith "Invalid input"
```
As @dinosaure said, maybe use something like `(t, [ `Msg of string ]) result` instead here.

bbenne10 replied:
In my unpushed changes, I've got further changes that switch to using results over throwing exceptions. 
I'm having one small issue with a utility function that is failing to compile, but I'll get that fixed (and hopefully pushed) today

pascutto commented on src/bloomf.ml:
```diff
@@ -78,6 +78,75 @@ let size_estimate t =
+  let rs = String.make 1 (Char.chr 30) in
+  let us = String.make 1 (Char.chr 31) in
+  let buf =
+    Buffer.create (5 + (4 * List.length t.p_len) + (2 * Bitv.length t.b))
```
How did you compute that size ? (also applies to other locations)

bbenne10 replied:
This is a lower bound. According to the docs, the buffer will get expanded as necessary. This simply assumes that each element of an array is one byte and multiplies by the length of each array + the number of separators we're adding. Getting this "wrong" isn't a huge deal, but I wanted to start somewhere that wasn't zero.

pascutto commented on src/bloomf.ml:
```diff
@@ -78,6 +78,75 @@ let size_estimate t =
+      string_of_int t.m;
+    ]
+  in
+  contents |> List.iter (fun s -> Buffer.add_string buf s);
```
```suggestion
  contents |> List.iter (Buffer.add_string buf);
```

bbenne10 replied:
Thanks for pointing this out!

pascutto commented on src/bloomf.ml:
```diff
@@ -78,6 +78,75 @@ let size_estimate t =
+  p_len_str
+  |> String.split_on_char rs
+  |> List.map (fun p ->
+         p |> String.split_on_char us |> Array.of_list |> parse_pair_ary)
```
I don't think you need to translate to an array, you can make `parse_pair_ary` to take a list instead.

pascutto reviewed (commented):
Thanks for your proposal, here is a first round of remarks for improvements (and possibly a bug fix too).

> You'll notice that I didn't use Bitv.output_bin.
> It only accepts channels and creating a memory-backed channel seemed quite daunting.

Ah ok I understand that. I'll check if we can add something like `{output,input}_bin_string` to `bitv` directly. In the meantime, I would suggest using the `L.{of,to}_string` functions instead (although they are pretty inefficient).

> Is the failing stage a lack of changes to CHANGES.md?
> If someone cares to explain the process (if there is one), I can also bump the version if desired.

You indeed need to add an entry to `CHANGES.md`, under an `## Unreleased` title.
No need to bump the version anywhere, that will be done during the next release.

> Ah - the failure is building under OCaml 4.03, as I'm using String.split_on_char which was introduced in 4.04.

I think it's fine to drop the support to 4.03 then. You can change that in the `bloomf.opam` file at the root of the project.
There is no global policy for depreciating versions of OCaml, it's up to each library to determine which versions they want to support.

> The other failure looked to be a cancelled build somewhere?

That's an unrelated CI failure, nothing to worry about.

> I also note that we rely on Random.self_init. Is there a reason for this? Because of this, I cannot reliably test that to_string returns a known good value. I can only test that it can read the values that it writes.

It is used in the tests to generate random data to fill the bloom filters and run tests on filled structures.
I support @dinosaure idea to fuzz test it using crowbar, but that can be left for another PR. I think simple tests could be added to the existing alcotest ones, checking that `x = of_string (to_string x))` for filled (or empty) filters.
I'm not sure I understand why the fact that the tests are randomized prevent you from testing the correction of the functions though?

bbenne10 commented:
Firstly: this is great feedback! Thank you!

> In the meantime, I would suggest using the `L.{of,to}_string` functions instead (although they are pretty inefficient).
Inefficient isn't a huge problem, I don't think. These operations should be relatively rare. 

I did, however, completely miss these functions. These will save quite a bit of parsing in our code. Thanks for pointing them out.

> It is used in the tests to generate random data to fill the bloom filters and run tests on filled structures.
I support @dinosaure idea to fuzz test it using crowbar, but that can be left for another PR. I think simple tests could be added to the existing alcotest ones, checking that x = of_string (to_string x)) for filled (or empty) filters.
I'm not sure I understand why the fact that the tests are randomized prevent you from testing the correction of the functions though?

My original idea had been to test `to_{string,bytes}` and `of_{string,bytes}` independently of one another.
It seemed that using `create_and_fill` would be the best way to do it, but being fully random I couldn't apply a fixed string/Bloom filter to test against. Thinking about it more, I have two options: Not using `create_and_fill` or just testing a full roundtrip. 

In my current, unpushed changes, I'm doing the latter. 

Speaking of unpushed changes: I'm still working on them. I discovered a few small serialization problems which I'm fixing (namely that mapping over an empty list returned an empty string, which wasn't then parsable and so needed a special case for both `p_len` and `b`). I should have the code updated before long - hopefully with most or all of the necessary changes.

pascutto commented:
> Inefficient isn't a huge problem, I don't think. These operations should be relatively rare.

They are inefficient in space: the bitvector `1;0;1;0;0;0;1;0;1;0` will be encoded as `"1010001010"` which is quite bad because you could encode 8 bits per char instead of one (that's what `output_bin` does). In the context of serialisation we want to avoid that, I'll keep you updated about what we can possibly add in `bitv` directly.

> My original idea had been to test `to_{string,bytes}` and `of_{string,bytes}` independently of one another.

We don't really care what the output of `to_string` is, as long as `of_string` gets you the correct filter back, so I would not bother with having fixed representations, and test that they are mutually inverse only, it's also more robust if the serialisation format changes later.

bbenne10 committed:
Use Bitv.L.{of,to}_string rather than handrolling serializers for Bitv.t

bbenne10 committed:
Add initial test for serialization and deserialization

bbenne10 committed:
Add documentation to {to,of}_{string,bytes}

bbenne10 committed:
Up minimum OCaml version to 4.04.0 for String.split_on_char

bbenne10 committed:
Add changes to CHANGES.md

bbenne10 commented:
Sorry for the churn here lately - I think this should finalize my changes as I see them so far.
Everything seems to work in my testing via the top-level and tests.
I have switched to a `Stdlib.result` oriented workflow, tests, added documentation to the interface, updated the CHANGES file and have updated the minimum OCaml version. 
I have not added the dependency on crowbar.

There are probably nitpicks about style and there are likely bugs or tweaks you all would like to see made. 
Please don't hesitate to ask for them :)

Regarding inefficiency of space: There's a lot we could likely do here, such as converting to a series of hexadecimal or base-10 integers or ASCII characters which would reduce space considerably (at the cost of additional complexity due to padding etc). Please let me know if there's anything you'd like me to do on this front.

EDITED to add: For all of that churn, the history on this branch is a bit misleading. I accidentally added the `results` changes in with the switch to `Bitv.L` based parsing. if you'd like that squashed or separated out, I'm happy to do it but hesitant to churn much more without a go-ahead on it. Again: let me know.

bbenne10 committed:
Small style fixup

Co-authored-by: Clément Pascutto <clement@pascutto.fr>

bbenne10 committed:
Avoid converting to Array just to match pair

bbenne10 commented:
Ping? Checking in on this. Has anyone had any time to look it over?

pascutto commented:
I'm figuring out the serialisation and release on the `bitv` end, will come back to you later this week :)

bbenne10 commented:
In light of a lack of progress on this, I'm working on an implementation that converts the bit vector to ascii characters.
I'm fighting with padding right now on non mod-8 input, but it should be pretty simple otherwise.
I'm happy to push this implementation here, if you'd like.


EDIT: Okay - I am confused. There's a new release of `backtracking/bitv` that contains changes from @pascutto that expose `{to,of}_bytes`. However, these changes aren't included in the "latest release" on github (which is still 1.3?), nor do the changes show up on opam. 

What is going on?

pascutto commented:
I indeed added `{to,of}_bytes` in `bitv`, but there is currently an issue with `dune-release` that prevents the release for some reason. This is being investigated.
In the meantime, you can pin the current state of the repo to still be able to use it here.
Then a couple more tests would be welcome (e.g. with a larger bitv for instance), then we're good to merge.
Sorry this is taking time, I'm also busy with other projects :)

bbenne10 commented:
I'll take a crack at adding additional tests today.

pascutto commented:
We released `bitv.1.4` just now (https://github.com/ocaml/opam-repository/pull/17240), We just have to wait until it's merged and you can use it here.

bbenne10 commented:
Because `Bitv.{to,of}_bytes` is (obviously) bytes-native, the current implementation won't work. The separator character between "sections" is now a problem, as arbitrary byte streams make it harder to fix a delineation point. I was using ascii characters intended for this, but that's less feasible now, since the byte stream for arbitrary `Bitv.t` may include those bytes unintentionally. 

I'll get around to this soon, but it may be this weekend.

bbenne10 committed:
Pin to Bitv 1.4+

bbenne10 committed:
Switch to bytes native serialize

Also switches to Length / Value based encoding scheme.

bbenne10 commented:
Just removed the `to_string` and `of_string` and switched to a length/value encoding scheme (which makes the parsing a bit easier to do recursively).

pascutto commented on CHANGES.md:
```diff
@@ -1,3 +1,7 @@
+## Unreleased 2020-08-26
+
+- Add serializers and deserializers: `to_string`, `to_bytes`, `of_string`, and `of_bytes`
```
```suggestion
- Add `to_bytes` and `of_bytes`
```

pascutto commented on CHANGES.md:
```diff
@@ -1,3 +1,7 @@
+## Unreleased 2020-08-26
```
```suggestion
## Unreleased
```

pascutto reviewed (commented):
Thank you for your work! A couple of last minutes remarks and I can merge it.
- Could you format the code with `dune build @fmt --auto-promote` to make the CI happy?
- I think the implementation could be greatly simplified by adopting the same `(length, data)` convention for `m`, `k` and `p_len`; but I can do that in another PR.
- Similarly, that should work without having to allocate many buffers and bytes, but I can fix that later too.

bbenne10 committed:
Update CHANGES.md to reflect new interface (and remove date)

bbenne10 commented:
> Thank you for your work! A couple of last minutes remarks and I can merge it.

Of course! I'm glad to help :)

> * Could you format the code with `dune build @fmt --auto-promote` to make the CI happy?

I did and there's no local changes. I *think* the failures have to do with `lint-fmt` being dependent on ocamlformat 0.14.2, which is explicitly not compatible with OCaml 4.11.0 (which it implicitly seems to build with). I had this problem when linting locally and just rolled back to OCaml 4.10.0. We could upgrade to latest ocamlformat (currently 0.15) and resolve the issue that way. Is that acceptable?

> * I think the implementation could be greatly simplified by adopting the same `(length, data)` convention for `m`, `k` and `p_len`; but I can do that in another PR.

I agree that it is possible, but some care needs to be taken here. I thought about doing this, but didn't think it'd be easier. `m` and `k` already use it, but `p_len` (though it does use L/V encoding for writing the top-level bit) is problematic because it's a simple int list. Unless I am misunderstanding your wishes, we could either adopt a fixed width for each element or we would need a different separator for the `p_len` list and we'd lose the benefit of size reduction because we'd need to specify the size of each list element. (that is - unless we can make the assumption that each element in `p_len` is {n} bytes, which would make this whole thing much easier).

I'd love to support that case, but I wasn't sure of the assumptions I could safely make about `p_len`'s contents.

>* Similarly, that should work without having to allocate many buffers and bytes, but I can fix that later too.

I'd love to do that too! But I just couldn't figure out a relatively performant way to do so without Buffers. I didn't want to re-allocate strings every time and I couldn't make a safe bet about how large the Bytes object needed to be ahead of time. 

If you have thoughts here, please let me know. This is largely a learning process for me and I greatly appreciate the feedback you have given me and genuinely would love to learn more. 

Edit: Fixed formatting

pascutto commented:
> I _think_ the failures have to do with `lint-fmt` being dependent on ocamlformat 0.14.2, which is explicitly not compatible with OCaml 4.11.0 (which it implicitly seems to build with). I had this problem when linting locally and just rolled back to OCaml 4.10.0. We could upgrade to latest ocamlformat (currently 0.15) and resolve the issue that way. Is that acceptable?

Ah that's right, sorry! Yes please upgrade to ocamlformat.0.15.0 in `.ocamlformat`.

>  we could either adopt a fixed width for each element or we would need a different separator for the `p_len` list and we'd lose the benefit of size reduction because we'd need to specify the size of each list element.

I think the easiest way to go is to use fixed-length for all `int`s (8 bytes), independently of the machine/version of OCaml (look how it's done [in bitv](https://github.com/backtracking/bitv/blob/master/bitv.ml#L542-L551) for instance).
Once you have that, encoding the bloomf is making a `Bytes.t` of size `8 (* m *) + 8 (* k *) + (16 * k) (* p_len *) + (Bytes.length encoded_b)`, then blitting (or better, just writing them directly to avoid intermediate allocations) them in order.

> I'd love to do that too! But I just couldn't figure out a relatively performant way to do so without Buffers. I didn't want to re-allocate strings every time and I couldn't make a safe bet about how large the Bytes object needed to be ahead of time.

I think the fixed length approach solves that issue :)

What do you think?

bbenne10 committed:
Update ocamlformat version to 0.15.0

bbenne10 commented:
Sounds perfect. Taking a stab as I type this :)

bbenne10 commented:
Open source contributions have taken a back seat lately, but I have an implementation that seems to work okay. Drastically less code and much simpler! I'm just going to go back and add some error checking (and hopefully turn the exception based parsing into `result`s, if it makes sense) 

Sorry for the delay here, but I'll get to it!

bbenne10 committed:
Update {to,of}_bytes to use Bitv's bytes serialization implementation

Much cleaner interface, many fewer allocations!

bbenne10 commented:
Okay - I've pushed everything I have so far.
Tests pass locally - but I am seeing failures on arm32 and ppc64 in the CI builds, which I have no idea how to debug.
I wonder if it has to do with the int size on these platforms?

bbenne10 commented:
I just realized that I never followed up on this. Does anyone have any thoughts on the CI failures? Ive not tried reproducing the arch specific containers locally, but I am unsure that'll help, as I dont have machines with these two architectures.

dinosaure commented on src/bloomf.ml:
```diff
@@ -78,6 +78,50 @@ let size_estimate t =
+          parse_p_len (start + 8) stop ((value, next_value) :: acc) None
+      | None -> parse_p_len (start + 8) stop acc (Some next_value)
+  in
+  match int_of_bytes (Bytes.sub b 0 8) with
```
You assume here that the `integer` is a `int64` (8 bytes) but it does not work on 32-bits architecture. To solve that, you should use `Int64` and make something like the `int64_of_bytes` function.

dinosaure replied:
(May be for more context, an OCaml `int` on a 32 bits architecture is a 31-bits integer - on a 64-bits architecture, it's a 63-bits integer - you can easily simulate the situation with an OCaml switch `4.10.0+32bit` - `opam switch create 4.10.0+32bit`)

bbenne10 replied:
I knew the bit size of native integers was smaller on 32 bit systems, but the issue I see with this is that the x86_32 builds passed, which seems odd to me. I'll check this out on a 32bit switch (I've not had a lot of time for this in the recent weeks, but I should be able to test this in the next few days). 

Thank you for the review.

bbenne10 committed:
Switch to Int64 based workflow

bbenne10 commented:
Sorry for the noise - committed and pushed without thinking this would update here. I've finally found a few hours to work on this. Hope to have some meaningful progress before long.

bbenne10 commented:
I have converted everything to utilizing `Int64`. I am still seeing failures on arm32 and ppc64, but I'm unsure as to exactly why and I unfortunately don't have hardware to test this out on. It's late here and I've got to be up in the morning, but I intend to test this on virtual hardware as soon as I can get away from work + family for a few more hours.

This is working for x86_32 locally (and passes on CI as well), so again: I'm stumped.

pascutto commented:
Hi! I'm sorry I took such a long time to get back to you.
I have added simple serialization functions in d581f770a58f23531f2dea54d2c4f83d44d141a4, as I don't have enough time these days to help you go with this PR unfortunately :/
I have implemented most of your ideas so you are obviously credited and your contribution was very welcome.
If you feel like something else is missing please feel free to reopen an issue/PR.

bbenne10 commented:
No need for apologies. Life got busy (and work didn't need this functionality immediately - that all got held up in contracting for reasons unrelated to this feature) and I never got around to fixing this up. It's been on my TODO for a LONG time. Thanks for getting around to implementing it!
